core/services/vrf/v2/listener_v2_log_listener.go

```diff
@@ -0,0 +1,451 @@
+		}
+	}
+}
+
+// initializeLastProcessedBlock returns the earliest block number that we need to
+// process requests for. This is the block number of the earliest unfulfilled request
+// or the latest finalized block, if there are no unfulfilled requests.
+// TODO: add tests
+func (lsn *listenerV2) initializeLastProcessedBlock(ctx context.Context) (lastProcessedBlock int64, err error) {
+	lp := lsn.chain.LogPoller()
+	start := time.Now()
+
+	// will retry on error in the runLogListener loop
+	latestBlock, err := lp.LatestBlock()
+	if err != nil {
+		return 0, errors.Wrap(err, "LogPoller.LatestBlock()")
+	}
+	fromTimestamp := time.Now().UTC().Add(-lsn.job.VRFSpec.RequestTimeout)
+	ll := lsn.l.With(
+		"latestFinalizedBlock", latestBlock.FinalizedBlockNumber,
+		"latestBlock", latestBlock.BlockNumber,
+		"fromTimestamp", fromTimestamp)
+	ll.Debugw("Initializing last processed block")
+	defer func() {
+		ll.Debugw("Done initializing last processed block", "elapsed", time.Since(start))
+	}()
+
+	numBlocksToReplay := numReplayBlocks(lsn.job.VRFSpec.RequestTimeout, lsn.chain.ID())
+	ll.Debugw("running replay on log poller")
+	err = lp.Replay(ctx, mathutil.Max(latestBlock.FinalizedBlockNumber-numBlocksToReplay, 1))
+	if err != nil {
+		return 0, errors.Wrap(err, "LogPoller.Replay")
+	}
+
+	// get randomness requested logs with the appropriate keyhash
+	// keyhash is specified in topic1
+	requests, err := lp.IndexedLogsCreatedAfter(
+		lsn.coordinator.RandomWordsRequestedTopic(), // event sig
+		lsn.coordinator.Address(),                   // address
+		1,                                           // topic index
+		[]common.Hash{lsn.job.VRFSpec.PublicKey.MustHash()}, // topic values
+		fromTimestamp,       // from time
+		logpoller.Finalized, // confs
+	)
+	if err != nil {
+		return 0, errors.Wrap(err, "LogPoller.LogsCreatedAfter RandomWordsRequested logs")
+	}
+
+	// fulfillments don't have keyhash indexed, we'll have to get all of them
+	// TODO: can we instead write a single query that joins on request id's somehow?
+	fulfillments, err := lp.LogsCreatedAfter(
+		lsn.coordinator.RandomWordsFulfilledTopic(), // event sig
+		lsn.coordinator.Address(),                   // address
+		fromTimestamp,                               // from time
+		logpoller.Finalized,                         // confs
+	)
+	if err != nil {
+		return 0, errors.Wrap(err, "LogPoller.LogsCreatedAfter RandomWordsFulfilled logs")
+	}
+
+	unfulfilled, _, _ := lsn.getUnfulfilled(append(requests, fulfillments...), ll)
+	var earliestUnfulfilledBlock int64 = math.MaxInt64
+	for _, req := range unfulfilled {
+		if req.Raw().BlockNumber < uint64(earliestUnfulfilledBlock) {
+			earliestUnfulfilledBlock = int64(req.Raw().BlockNumber)
+		}
+	}
+	if earliestUnfulfilledBlock == math.MaxInt64 {
+		// no unfulfilled requests
+		return latestBlock.FinalizedBlockNumber, nil
+	}
+
+	// earliestUnfulfilledBlock is <= latestBlock.FinalizedBlockNumber because in our queries we specify
```

Comment:
given this, can the above logic be simplified a bit like ~

earliestUnfilfilledBlock := latest.FinalizedBlockNumber
for _,req := range unfulfilled {
	if req.Raw.BlockNumber < earliestUnfilledBlock {
		//update earliest
}
}

in words: it seems impossible for line 163 to be true unless len(unfilled) == 0 and if every elem of unfilled is by defn < finalizedblocknumber, then you can get rid of the if on 163 by init'ing to the max value of finalizedblocknum

Reply:
Good catch, yeah I can simplify this probably